### PR TITLE
New Data Quality validation module to enable users to define data quality specifications

### PR DESCRIPTION
**Pull request by keenkeystrokes**
The new Data quality module enables users to define data quality specifications and provide an execution strategy.

#### What type of PR is this?
- Improvement

#### What does this PR do / why is it needed ?
It enables users to define data quality specifications and also provides methods to allow execution.

1. Introduces new grammar (parser, compiler, composer) to allow users to define data quality validations required.
2. Provides a function to generate custom lambda that can be used to create an execution plan 
3. Build artifact extension to allow the execution to be pre-generated as part of the project build

Co-authored-by: @SaitejaGovardhana

### CODE REVIEW

File: legend-engine-xts-dataquality/legend-engine-xt-dataquality-api/pom.xml [hunk 1]
```diff
@@ -0,0 +1,275 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>legend-engine-xts-dataquality</artifactId>
+        <groupId>org.finos.legend.engine</groupId>
+        <version>4.48.1-SNAPSHOT</version>
```
Comment (aormerod-gs) [inline legend-engine-xts-dataquality/legend-engine-xt-dataquality-api/pom.xml:8]: should this be 

```suggestion
        <version>4.48.2-SNAPSHOT</version>
```

File: legend-engine-xts-dataquality/legend-engine-xt-dataquality-compiler/pom.xml [hunk 2]
```diff
@@ -0,0 +1,162 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>legend-engine-xts-dataquality</artifactId>
+        <groupId>org.finos.legend.engine</groupId>
+        <version>4.48.1-SNAPSHOT</version>
```
Comment (aormerod-gs) [inline legend-engine-xts-dataquality/legend-engine-xt-dataquality-compiler/pom.xml:8]:
```suggestion
        <version>4.48.2-SNAPSHOT</version>
```

File: legend-engine-xts-dataquality/legend-engine-xt-dataquality-grammar/pom.xml [hunk 3]
```diff
@@ -0,0 +1,145 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>legend-engine-xts-dataquality</artifactId>
+        <groupId>org.finos.legend.engine</groupId>
+        <version>4.48.1-SNAPSHOT</version>
```
Comment (aormerod-gs) [inline legend-engine-xts-dataquality/legend-engine-xt-dataquality-grammar/pom.xml:8]:
```suggestion
        <version>4.48.2-SNAPSHOT</version>
```

File: legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/pom.xml [hunk 4]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>legend-engine-xts-dataquality</artifactId>
+        <groupId>org.finos.legend.engine</groupId>
+        <version>4.48.1-SNAPSHOT</version>
```
Comment (aormerod-gs) [inline legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/pom.xml:8]:
```suggestion
        <version>4.48.2-SNAPSHOT</version>
```

File: legend-engine-xts-dataquality/legend-engine-xt-dataquality-pure/pom.xml [hunk 5]
```diff
@@ -0,0 +1,298 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>legend-engine-xts-dataquality</artifactId>
+        <groupId>org.finos.legend.engine</groupId>
+        <version>4.48.1-SNAPSHOT</version>
```
Comment (aormerod-gs) [inline legend-engine-xts-dataquality/legend-engine-xt-dataquality-pure/pom.xml:8]:
```suggestion
        <version>4.48.2-SNAPSHOT</version>
```

File: legend-engine-xts-dataquality/pom.xml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>legend-engine</artifactId>
+        <groupId>org.finos.legend.engine</groupId>
+        <version>4.48.1-SNAPSHOT</version>
```
Comment (aormerod-gs) [inline legend-engine-xts-dataquality/pom.xml:8]:
```suggestion
        <version>4.48.2-SNAPSHOT</version>
```

File: legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/src/main/java/org/finos/legend/engine/protocol/dataquality/metamodel/DataSpaceDataQualityExecutionContext.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+// limitations under the License.
+
+
+package org.finos.legend.engine.protocol.dataquality.metamodel;
+
+import org.finos.legend.engine.protocol.pure.v1.model.context.PackageableElementPointer;
+
+public class DataSpaceDataQualityExecutionContext extends DataQualityExecutionContext
+{
+    public String _type = "dataSpaceDataQualityExecutionContext";
```
Comment (MauricioUyaguari) [inline legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/src/main/java/org/finos/legend/engine/protocol/dataquality/metamodel/DataSpaceDataQualityExecutionContext.java:22]: we shouldnt really carry this in the class. This should just be used to serialize/deseralize to json.

File: legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/src/main/java/org/finos/legend/engine/protocol/dataquality/model/DataQualityExecuteInput.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import org.finos.legend.engine.protocol.pure.v1.model.context.AlloySDLC;
+import org.finos.legend.engine.protocol.pure.v1.model.packageableElement.domain.ParameterValue;
+
+import java.util.List;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class DataQualityExecuteInput
+{
+    public AlloySDLC alloySDLC;
```
Comment (MauricioUyaguari) [inline legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/src/main/java/org/finos/legend/engine/protocol/dataquality/model/DataQualityExecuteInput.java:26]: Assuming this represents the model in the DQ Input. 
Does it make sense for this to be `PureModelContext` instead so we can run these checks both with grammar inputs as well as a full pmcd and not just pointed to alloy projects ?

**Review by MauricioUyaguari — COMMENTED** [review]
Was a bit curious on the model for this since I saw the studio changes. So left some small comments/questions. 
I would remove `_type` as a direct property from all the pojo classes. The jackson subtype annotations should be enough to handle serailize/deseralize of the class to json.

File: legend-engine-xts-dataquality/legend-engine-xt-dataquality-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataQualityCompilerExtension.java [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    /**
+     * Recursively go through hierarchical/generalization chain and find the property.
+     */
+    public static FastList<Constraint> resolveNodeConstraints(CompileContext context, org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.type.Class<?> _class, List<String> treeConstraintNames)
+    {
+        FastList<Constraint> constraints = FastList.newList();
+        for (CoreInstance c_type : _Class.computeConstraintsInHierarchy(_class, context.pureModel.getExecutionSupport().getProcessorSupport()))
+        {
+            Root_meta_pure_metamodel_constraint_Constraint_Impl constraint = (Root_meta_pure_metamodel_constraint_Constraint_Impl) c_type;
+            if (treeConstraintNames.contains(constraint._name)) {
```
Comment (aormerod-gs) [inline legend-engine-xts-dataquality/legend-engine-xt-dataquality-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataQualityCompilerExtension.java:236]:
```suggestion
            if (treeConstraintNames.contains(constraint._name))
            {
```

File: legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/src/main/java/org/finos/legend/engine/protocol/dataquality/model/DataQualityExecuteInput.java (hunk 8 above)
Comment (keenkeystrokes) [reply legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/src/main/java/org/finos/legend/engine/protocol/dataquality/model/DataQualityExecuteInput.java:26]: hi Mauricio, we do have an API which takes in a full pmcd, generates and execute the plan from it to pull defects. The request class is [DataQualityExecuteTrialInput](https://github.com/goldmansachs/legend-engine/blob/review-new-data-quality-module/legend-engine-xts-dataquality/legend-engine-xt-dataquality-api/src/main/java/org/finos/legend/engine/language/dataquality/api/DataQualityExecuteTrialInput.java) We will rename these classes better to DataQualityExecuteArtifactsInput and DataQualityExecuteInput. 

1. /execute - this API will be integrated in the studio where user can run the validation and fetch defects 
2. /executeArtifacts - this API will take in a reference to a project and then execute the stored plan to pull defects.(the above mentioned class in an input to this API)

File: legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/src/main/java/org/finos/legend/engine/protocol/dataquality/metamodel/DataSpaceDataQualityExecutionContext.java (hunk 7 above)
Comment (keenkeystrokes) [reply legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/src/main/java/org/finos/legend/engine/protocol/dataquality/metamodel/DataSpaceDataQualityExecutionContext.java:22]: _type is the property used in json subtypes for serialization and deserialization [DataQualityExecutionContext](https://github.com/goldmansachs/legend-engine/blob/review-new-data-quality-module/legend-engine-xts-dataquality/legend-engine-xt-dataquality-protocol/src/main/java/org/finos/legend/engine/protocol/dataquality/metamodel/DataQualityExecutionContext.java#L21). I think its needed here

this is similar to [RootGraphFetchTree](https://github.com/goldmansachs/legend-engine/blob/af6f1f9522b9ea36fbf388b447bd20295a756422/legend-engine-core/legend-engine-core-base/legend-engine-core-language-pure/legend-engine-protocol-pure/src/main/java/org/finos/legend/engine/protocol/pure/v1/model/valueSpecification/raw/classInstance/graph/RootGraphFetchTree.java#L21)

**Review by aormerod-gs — APPROVED** [review]

**Pull request merged**